staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go

```diff
@@ -278,59 +253,104 @@ func (p *oidcAuthProvider) idToken() (string, error) {
 		return "", fmt.Errorf("could not perist new tokens: %v", err)
 	}
-
-	// Update the in memory config to reflect the on disk one.
 	p.cfg = newCfg
 
-	return tokens.IDToken, nil
-}
-
-// oidcClient is the real implementation of the OIDCClient interface, which is
-// used for testing.
-type oidcClient struct {
-	client *oidc.Client
+	return idToken, nil
 }
 
-func (o *oidcClient) refreshToken(rt string) (oauth2.TokenResponse, error) {
-	oac, err := o.client.OAuthClient()
+// tokenEndpoint uses OpenID Connect discovery to determine the OAuth2 token
+// endpoint for the provider, the endpoint the client will use the refresh
+// token against.
+func tokenEndpoint(client *http.Client, issuer string) (string, error) {
+	// Well known URL for getting OpenID Connect metadata.
+	//
+	// https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig
+	wellKnown := strings.TrimSuffix(issuer, "/") + "/.well-known/openid-configuration"
+	resp, err := client.Get(wellKnown)
 	if err != nil {
-		return oauth2.TokenResponse{}, err
+		return "", err
 	}
+	defer resp.Body.Close()
 
-	return oac.RequestToken(oauth2.GrantTypeRefreshToken, rt)
-}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("oidc: failed to query metadata endpoint %s: %s", resp.Status, body)
+	}
 
-func (o *oidcClient) verifyJWT(jwt *jose.JWT) error {
-	return o.client.VerifyJWT(*jwt)
+	// Metadata object. We only care about the token_endpoint, the thing endpoint
+	// we'll be refreshing against.
+	//
+	// https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
+	var metadata struct {
+		TokenURL string `json:"token_endpoint"`
+	}
+	if err := json.Unmarshal(body, &metadata); err != nil {
+		return "", fmt.Errorf("oidc: failed to decode provider discovery object: %v", err)
+	}
+	if metadata.TokenURL == "" {
+		return "", fmt.Errorf("oidc: discovery object doesn't contain a token_endpoint")
+	}
+	return metadata.TokenURL, nil
 }
 
-func verifyJWTExpiry(now time.Time, s string) (valid bool, err error) {
-	jwt, err := jose.ParseJWT(s)
-	if err != nil {
-		return false, fmt.Errorf("invalid %q", cfgIDToken)
+func idTokenExpired(idToken string) (bool, error) {
+	parts := strings.Split(idToken, ".")
+	if len(parts) != 3 {
+		return false, fmt.Errorf("ID Token is not a valid JWT")
 	}
-	claims, err := jwt.Claims()
+
+	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
 	if err != nil {
 		return false, err
 	}
+	var claims struct {
+		Expiry jsonTime `json:"exp"`
+	}
+	if err := json.Unmarshal(payload, &claims); err != nil {
+		return false, fmt.Errorf("parsing claims: %v", err)
+	}
+
+	return time.Now().Add(expiryDelta).Before(time.Time(claims.Expiry)), nil
+}
+
+// jsonTime is a json.Unmarshaler that parses a unix timestamp.
+// Because JSON numbers don't differentiate between ints and floats,
+// we want to ensure we can parse either.
+type jsonTime time.Time
```

Comment:
An old version of dex used to do this because of handling claims as `map[string]interface{}` :( We've "fixed" it for all relevant but I've always done this defensively so if other providers do it I wouldn't know. golang.org/x/oauth2 does something similar for numeric values (expires_in)[1].

[1] https://github.com/golang/oauth2/blob/ad516a297a9f2a74ecc244861b298c94bdd28b9d/internal/token.go#L62-L63